docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1434,13 +1434,117 @@ The `ValueSet` message defines the following fields:
 * `preamble`, a `Preamble` message with the ID, name, and alias of this Value
   Set.
 
-* `bitwidth`, an `int32` set to the size in bits of the value matched in the
-  parser state.
+* `match`, a repeated field of `MatchField` messages, representing the list of
+  matches performed when looking up an expression in a Value Set. This
+  determines the format of the members which can be inserted into the Value Set
+  by the control plane, similarly to the `match_fields` repeated field in the
+  `Table` message.
 
 * `size`, an int32 representing the maximum number of entries (values) in the
   Value Set. It corresponds to the value of the size argument of the P4
   `value_set` constructor call.
 
+According to the P4 specification, the type parameter of a Value Set, which
+defines the type of the expression that can be matched against the Value Set in
+a parser transition, and therefore determines the format of the members that can
+be inserted into the Value Set by the control plane, must be one of `bit<W>`,
+`tuple`, or `struct` [@P4SelectExpr]. The rest of this section looks at all 3 of
+these cases and gives an example `ValueSet` message when appropriate.
+
+1. If the type parameter is `bit<W>`, `match` will include exactly one
+   `MatchField` message, with the following fields (if a field is omitted here,
+   it means the default Protobuf value should be used):
+
+ * `id`: set to 1
+ * `bitwidth`: set to the value of `W`
+ * `match_type`: set to `EXACT`
+
+~ Begin P4Example
+@id(1) value_set<bit<8> >(4) pvs;
+select (hdr.f8) { /* ... */ }
+~ End P4Example
+~ Begin Prototext
+value_sets {
+  preamble {
+    id: 0x03000001
+    name: "pvs"
+  }
+  match {
+    id: 1
+    bitwidth: 8
+    match_type: EXACT
+  }
+  size: 4
+}
+~ End Prototext
+
+2. If the type parameter is a `tuple`, this version of P4Runtime does not
+   support runtime programming of the Value Set. If the P4Info message is
+   generated by a compiler, and the P4 program includes such a Value Set, the
+   compiler must reject the program.
+
+3. If the type parameter is a `struct`, this version of P4Runtime requires that
+   all the fields of the struct be of type `bit<W>` (where `W` can be different
+   for each field). If the P4Info message is generated by a compiler, and the P4
+   program includes such a Value Set, the compiler must reject the
+   program. If the Value Set is supported, the `match` field will include one
+   `MatchField` message for each field in the struct, with the following fields:
+
+ * `id`: must be unique with respect to the other `match` entries. If the P4Info
+   message was generated from a P4 compiler, we recommend that the IDs be
+   assigned incrementally, starting from 1, in the same order as the fields in
+   the P4 struct declaration.
+ * `name`: set to the name of the corresponding struct field.
+ * `annotations`: set to the list of P4 annotations associated with the struct
+   field.
```

Comment:
Maybe say something like this here?   "except for the `@match` annotation, if present -- see the `match` field below."

Reply:
Done